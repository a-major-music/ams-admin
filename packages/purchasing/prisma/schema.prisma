// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
  output   = "./generated/client"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Address {
  id         Int       @id @default(autoincrement())
  guid       String    @unique @default(uuid())
  address1   String
  address2   String?
  address3   String?
  city       String?
  postcode   String?
  supplier   Supplier? @relation(fields: [supplierId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  supplierId Int       @unique
}

model Contact {
  id         Int      @id @default(autoincrement())
  guid       String   @unique @default(uuid())
  name       String
  email      String?
  phone      String?
  supplier   Supplier @relation(fields: [supplierId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  supplierId Int      @unique
}

model Supplier {
  id             Int             @id @default(autoincrement())
  guid           String          @unique @default(uuid())
  name           String          @unique
  website        String?
  address        Address?
  contacts       Contact[]
  purchaseOrders PurchaseOrder[]
}

model PurchaseOrder {
  id          Int                     @id @default(autoincrement())
  guid        String                  @unique @default(uuid())
  number      String                  @unique
  state       String                  @default("draft")
  issueDate   DateTime?
  pendingXero Boolean                 @default(true)
  lineItems   PurchaseOrderLineItem[]
  subTotal    Int
  taxAmount   Int
  total       Int
  supplier    Supplier?               @relation(fields: [supplierId], references: [id])
  supplierId  Int?
}

model PurchaseOrderLineItem {
  id                  Int                            @id @default(autoincrement())
  guid                String                         @unique @default(uuid())
  variantSku          String?
  variantGuid         String?
  variantUPC          String?
  variantSupplierCode String?
  variantTaxRate      Int
  description         String
  variantMediaURI     String?
  itemCost            Int
  quantityOrdered     Int
  packSize            Int @default(1)
  receipts            PurchaseOrderLineItemReceipt[]
  purchaseOrder       PurchaseOrder                  @relation(fields: [purchaseOrderId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  purchaseOrderId     Int
}

model PurchaseOrderLineItemReceipt {
  id                      Int                    @id @default(autoincrement())
  guid                    String                 @unique @default(uuid())
  date                    DateTime               @default(now())
  received                Int
  receivedPrice           Int
  PurchaseOrderLineItem   PurchaseOrderLineItem? @relation(fields: [purchaseOrderLineItemId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  purchaseOrderLineItemId Int?
}
