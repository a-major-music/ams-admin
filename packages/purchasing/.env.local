# Local development environment for Purchasing Service
DATABASE_URL="postgresql://postgres:postgres@localhost:5432/purchasing?schema=public"
PORT=3001

# Auth0 Configuration
AUTH0_DOMAIN=amajormusic.us.auth0.com
AUTH0_AUDIENCE=https://api.amajormusic.co.uk

# Optional: Add your local environment variables here
# XERO_API_KEY=your_key_here