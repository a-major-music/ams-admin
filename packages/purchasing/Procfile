web: yarn workspace @amm/amm-doc build && yarn workspace @amm/apikey build && yarn workspace @amm/purchasing start
