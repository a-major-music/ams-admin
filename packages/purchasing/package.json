{
  "name": "@amm/purchasing",
  "version": "1.0.0",
  "engines": {
    "node": ">=20.x",
    "npm": ">=10.0.0"
  },
  "description": "Manage Purchase Orders and Suppliers",
  "main": "main.ts",
  "private": true,
  "scripts": {
    "prebuild": "rimraf dist",
    "build": "nest build",
    "clean": "npx prisma db push --force-reset",
    "prisma:migrate": "npx prisma migrate deploy",
    "prisma:generate": "npx prisma generate",
    "format": "prettier --write \"src/**/*.ts\" \"test/**/*.ts\"",
    "start": "npm run prisma:migrate && npm run prisma:generate && nest start",
    "start:dev": "npm run prisma:generate && nest start --watch",
    "start:debug": "npm run prisma:generate && nest start --debug --watch",
    "start:prod": "npm run prisma:migrate && npm run prisma:generate && node dist/main.js",
    "lint": "eslint \"{src,apps,libs,test}/**/*.ts\" --fix",
    "test": "jest",
    "test:watch": "jest --watch",
    "test:cov": "jest --coverage",
    "test:debug": "node --inspect-brk -r tsconfig-paths/register -r ts-node/register node_modules/.bin/jest --runInBand",
    "test:e2e": "export $(cat .env-e2e | xargs) && npx prisma db push --force-reset && jest --config ./test/jest-e2e.json --detectOpenHandles"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@amm/amm-doc": "file:../amm-doc",
    "@amm/types": "file:../types",
    "@nestjs/axios": "^3.0.3",
    "@nestjs/cli": "^11.0.7",
    "@nestjs/common": "^11.1.5",
    "@nestjs/config": "^4.0.2",
    "@nestjs/core": "^11.1.5",
    "@nestjs/event-emitter": "^2.0.4",
    "@nestjs/platform-express": "^11.1.5",
    "@nestjs/swagger": "^11.2.0",
    "@prisma/client": "^6.12.0",
    "@types/jest": "^29.5.5",
    "pg": "^8.7.3",
    "ts-jest": "^29.1.1"
  },
  "jest": {
    "moduleFileExtensions": [
      "js",
      "json",
      "ts"
    ],
    "rootDir": "src",
    "testRegex": ".*\\.spec\\.ts$",
    "transform": {
      "^.+\\.(t|j)s$": "ts-jest"
    },
    "collectCoverageFrom": [
      "**/*.(t|j)s"
    ],
    "coverageDirectory": "../coverage",
    "testEnvironment": "node"
  },
  "devDependencies": {
    "@golevelup/ts-jest": "^0.3.2",
    "@nestjs/testing": "^11.1.5",
    "@types/node": "^20.0.0",
    "@types/validator": "^13.6.5",
    "node-gyp": "^9.0.0",
    "prisma": "^6.12.0",
    "supertest": "^6.2.2"
  }
}
