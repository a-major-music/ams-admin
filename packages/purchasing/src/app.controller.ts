import { Res } from "@nestjs/common";
import { Controller, Get } from "@nestjs/common";
import { ApiTags, ApiOperation } from "@nestjs/swagger";
import { AppService } from "./app.service";

@ApiTags("Main")
@Controller()
export class AppController {
  constructor(private readonly appService: AppService) {}

  @Get()
  @ApiOperation({ description: "Loads this swagger page" })
  redirect(@Res() res) {
    return res.redirect("/swagger-ui");
  }
}
