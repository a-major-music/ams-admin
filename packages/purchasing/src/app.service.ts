import { Injectable } from '@nestjs/common';

@Injectable()
export class AppService {
  getHello(): string {
    return 'Hello World! <a href=/swagger-ui>swagger-ui</a>';
  }
}
