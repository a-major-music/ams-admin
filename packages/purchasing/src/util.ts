export function removeProps(obj: object, props: string[]): object {
  return JSON.parse(
    JSON.stringify(obj, (k, v) => (props.includes(k) ? undefined : v))
  );
}
