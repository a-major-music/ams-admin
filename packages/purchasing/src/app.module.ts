import { Module } from "@nestjs/common";
import { ConfigModule } from "@nestjs/config";
import { AppController } from "./app.controller";
import { AppService } from "./app.service";
import { PurchasingModule } from "./purchasing/purchasing.module";
import { SupplierModule } from "./supplier/supplier.module";
import { AmmDocModule } from "@amm/amm-doc";
import { EventEmitterModule } from "@nestjs/event-emitter";
import { AuthModule } from "../../common/src/auth/auth.module";

@Module({
  imports: [
    ConfigModule.forRoot(),
    AuthModule,
    PurchasingModule,
    SupplierModule,
    AmmDocModule,
    EventEmitterModule.forRoot(),
  ],
  controllers: [AppController],
  providers: [AppService],
})
export class AppModule {}
