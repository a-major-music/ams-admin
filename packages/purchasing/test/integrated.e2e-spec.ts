import { INestApplication } from "@nestjs/common";
import { Test } from "@nestjs/testing";
import { AppModule } from "../src/app.module";
import { PrismaService } from "../src/prisma.service";
import { PurchasingModule } from "../src/purchasing/purchasing.module";
import { SupplierModule } from "../src/supplier/supplier.module";
import supertest = require("supertest");
import { removeProps } from "../src/util";

let app: INestApplication;
let prisma: PrismaService;

beforeAll(async () => {
  const module = await Test.createTestingModule({
    imports: [AppModule, PurchasingModule, SupplierModule],
  }).compile();

  app = module.createNestApplication();
  prisma = app.get<PrismaService>(PrismaService);
  await app.init();
});

afterAll(async () => {
  await app.close();
});

describe("when creating a purchase order", () => {
  it("should create a new supplier and link the PO", async () => {
    const s = {
      name: "Testing create supplier",
      website: "https://nowhere.com",
      address: {
        address1: "Address 1",
        city: "City",
        postcode: "XXX",
      },
      contacts: [
        {
          name: "Testing create supplier (contact)",
        },
      ],
    };

    // Create a supplier
    const sR = await supertest
      .agent(app.getHttpServer())
      .post("/supplier/create")
      .set("Accept", "application/json")
      .send(s)
      .expect("Content-Type", /json/)
      .expect(201);

    const po = {
      number: "PO0001",
      state: "draft",
      lineItems: [
        {
          variantSku: "BOK00001",
          variantTaxRate: 20,
          itemCost: 1000,
          quantityOrdered: 10,
          description: "",
        },
      ],
      supplier: sR.body,
      subTotal: 10000,
      taxAmount: 2000,
      total: 12000,
    };

    const saved = await supertest
      .agent(app.getHttpServer())
      .post("/purchasing/create")
      .set("Accept", "application/json")
      .send(po)
      .expect("Content-Type", /json/)
      .expect(201);

    // Now get the PO back, by GUID.
    const testR = await supertest
      .agent(app.getHttpServer())
      .get(`/purchasing/${saved.body.guid}`)
      .set("Accept", "application/json")
      .expect("Content-Type", /json/)
      .expect(200);

    expect(removeProps(testR.body, ["guid"])).toMatchSnapshot();

    await prisma.purchaseOrderLineItem.deleteMany({});
    await prisma.purchaseOrder.deleteMany({});
    await prisma.supplier.deleteMany({});
  });

  // it("should have a supplier", () => {});
  // it("the supplier should have contact details", () => {});
});
