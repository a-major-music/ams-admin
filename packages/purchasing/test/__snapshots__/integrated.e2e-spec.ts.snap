// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when creating a purchase order should create a new supplier and link the PO 1`] = `
Object {
  "id": 4,
  "issueDate": null,
  "lineItems": Array [
    Object {
      "description": "",
      "id": 7,
      "itemCost": 1000,
      "purchaseOrderId": 4,
      "quantityOrdered": 10,
      "receipts": Array [],
      "variantGuid": null,
      "variantMediaURI": null,
      "variantSku": "BOK00001",
      "variantSupplierCode": null,
      "variantTaxRate": 20,
      "variantUPC": null,
    },
  ],
  "number": "PO0001",
  "pendingXero": true,
  "state": "draft",
  "subTotal": 10000,
  "supplier": Object {
    "address": Object {
      "address1": "Address 1",
      "address2": null,
      "address3": null,
      "city": "City",
      "id": 1,
      "postcode": "XXX",
      "supplierId": 3,
    },
    "contacts": Array [
      Object {
        "email": null,
        "id": 1,
        "name": "Testing create supplier (contact)",
        "phone": null,
        "supplierId": 3,
      },
    ],
    "id": 3,
    "name": "Testing create supplier",
    "website": "https://nowhere.com",
  },
  "supplierId": 3,
  "taxAmount": 2000,
  "total": 12000,
}
`;
